docs/testing.md

```diff
@@ -0,0 +1,340 @@
+
+For further reading, you may find these resources helpful:
+
+- [Jest Docs](https://jestjs.io/docs/getting-started)
+- [React Testing Library Docs](https://testing-library.com/docs/react-testing-library/intro/)
+- [Testing playground](https://testing-playground.com/) - use good testing practices to match elements.
+
+## Cypress
+
+Cypress is a test runner used for **End-to-End (E2E)** and **Integration Testing**.
+
+### Quickstart
+
+You can use `create-next-app` with the [with-cypress example](https://github.com/vercel/next.js/tree/canary/examples/with-jest) to quickly get started.
+
+```bash
+npx create-next-app --example with-cypress with-cypress-app
+```
+
+### Manual setup
+
+To get started with Cypress, install the `cypress` package:
+
+```bash
+npm install --save-dev cypress
+```
+
+Add Cypress to the `package.json` scripts field:
+
+```json
+"scripts": {
+	"dev": "next dev",
+  "build": "next build",
+  "start": "next start",
+	"cypress": "cypress open",
+}
+```
+
+Run Cypress for the first time to generate examples that use their recommended folder structure:
+
+```bash
+npm run cypress
+```
+
+You can look through the generated examples and the [Writing Your First Test](https://docs.cypress.io/guides/getting-started/writing-your-first-test) section of the Cypress Documentation to help you get familiar with Cypress.
+
+### Creating your first Cypress integration test
+
+Assuming the following two Next.js pages:
+
+```jsx
+// pages/index.js
+import Link from 'next/link'
+
+export default function Home() {
+  return (
+    <div>
+      <Link href="/about">
+        <a>About</a>
+      </Link>
+    </div>
+  )
+}
+```
+
+```jsx
+// pages/about.js
+import Link from 'next/link'
+
+export default function About() {
+  return (
+    <div>
+      <h1>About page</h1>
+    </div>
+  )
+}
+```
+
+Add a simple test to check your navigation is working correctly:
+
+```jsx
+// cypress/integration/app.spec.js
+
+describe('Navigation', () => {
+  it('should navigate to the about page', () => {
+    // Start from the index page
+    cy.visit('http://localhost:3000/')
+
+    // Find a link with an href attribute containing "about" and click it
+    cy.get('a[href*="about"]').click()
+
+    // The new url should include "/about"
+    cy.url().should('include', '/about')
+
+    // The new page should contain an h1 with "About page"
+    cy.get('h1').contains('About page')
+  })
+})
+```
+
+You can use `cy.visit("/")` instead of `cy.visit("http://localhost:3000/")` if you add `"baseUrl": "http://localhost:3000"` to the `cypress.json` configuration file.
+
+### Running your Cypress tests
+
+Because Cypress is testing a real Next.js application, it requires the Next.js server to be running prior to starting Cypress. Run `npm run dev` to start the development server then run `npm run cypress` in another terminal window to start Cypress.
```

Comment:
Is it better to run in dev mode or do `next build && next start`?

Reply:
Do you have a specific reason? I didn't see any examples using `next build`.

Reply:
I've seen some users on Twitter talk about issues with the dev server in these tests, and to more accurately simulate production, using a prod build.